File: src/services/hrr.ts
import { DB } from '../database';
import { useLocalCache } from '../util/cache/localCache';
import { useRedisCache } from '../util/cache/redisCache';
import { useCache, useData } from '../util/useData';

type GenericRule = {
    pattern: string;
};

export type Header = {
    source: string;
    headers: {
        [key: string]: string;
    }[];
    has: {
        type: 'header' | 'cookie' | 'host' | 'query';
        key: string;
        value?: string;
    }[];
};

type Condition = {
    type: 'header' | 'cookie' | 'host' | 'query';
    key: string;
    value?: string;
};

type HeaderRule = GenericRule & {
    // Conditions that need to be true in order for the rule to apply
    conditions: Condition[];
    // Headers to add when the rule is applied
    headers: Record<string, string>;
};

type RewriteRule = GenericRule & {
    // Conditions that need to be true in order for the rule to apply
    conditions: Condition[];
    destination: string;
};
type RedirectRule = GenericRule & {
    // Conditions that need to be true in order for the rule to apply
    conditions: Condition[];
    status: 301 | 302 | 307;
    destination: string;
};
type Rules = HeaderRule | RewriteRule | RedirectRule;

export const getHeaderRules = async (
    deploy_id: bigint,
    path: string
): Promise<HeaderRule[]> => {
    const rules = useData(`header_${deploy_id}`);

    return [];
};
export const getRedirectRules = async (
    deploy_id: bigint,
    path: string
): Promise<RedirectRule[] | undefined> =>
    await useCache<RedirectRule[]>(
        `redirect_${deploy_id}`,
        useLocalCache(),
        useRedisCache(),
        async (_key: string) => {
            const data = await DB.selectOneFrom(
                'deployment_configs',
                ['redirects'],
                { deploy_id }
            );

            if (!data) return;

            const { redirects } = data;

            return JSON.parse(redirects);
        }
    );

export const getRewriteRules = async (
    deploy_id: bigint,
    path: string
): Promise<RewriteRule[]> => {
    return [];
};

export const matchRedirects = (
    redirects: RedirectRule[],
    path: string
): RedirectRule | undefined => {
    for (const redirect of redirects) {
        if (new RegExp(redirect.pattern).test(path)) {
            return Object.assign(
                {},
                {
                    status: 302,
                },
                redirect
            );
        }
    }

    return undefined;
};
export const matchRewrites = (
    rewrites: RewriteRule[],
    path: string
): RewriteRule | undefined => {
    for (const rewrite of rewrites)
        if (new RegExp(rewrite.pattern).test(path)) return rewrite;

    return undefined;
};

File: apps/router/services/hrr.go
package services

import (
	"encoding/json"
	"fmt"
	"github.com/v3xlabs/edgeserver/apps/router/db"
	"github.com/v3xlabs/edgeserver/apps/router/models"
	"github.com/v3xlabs/edgeserver/apps/router/services/cache"
	"log"
	"regexp"
	"sync"
)

type GenericRule struct {
	Pattern string
}

type Condition struct {
	Type  string
	Key   string
	Value string
}

type HeaderRule struct {
	GenericRule
	Conditions []Condition
	Headers    map[string]string
}

type RewriteRule struct {
	GenericRule
	Conditions  []Condition
	Destination string
}

type RedirectRule struct {
	GenericRule
	Conditions  []Condition
	Status      int
	Destination string
}

// TODO: Implement Conditions

func getRedirectRules(deployId int64) (*[]RedirectRule, error) {
	return cache.UseCache(
		fmt.Sprintf("redirect_%d", deployId),
		[]cache.ResolverSetter[[]RedirectRule]{
			cache.UseLocalCache[[]RedirectRule](),
			cache.UseRedisCache[[]RedirectRule](),
			cache.ResolverSetter[[]RedirectRule]{
				Resolver: func(key string) (*[]RedirectRule, error) {
					session := db.Get()
					data := models.DeploymentConfigsStruct{
						DeployId: deployId,
					}

					q := session.Query(models.DeploymentConfigs.Get("redirects")).BindStruct(data)
					err := q.GetRelease(&data)
					log.Printf("getRedirectRulesDB: %v", data)

					result := make([]RedirectRule, 0)

					if err = json.Unmarshal([]byte(data.Redirects), &result); err != nil {
						return nil, err
					}

					return &result, err
				},
			},
		},
	)
}

func MatchRedirects(redirects *[]RedirectRule, path string) *RedirectRule {
	if redirects == nil {
		return nil
	}

	for _, redirect := range *redirects {
		match, err := regexp.MatchString(redirect.Pattern, path)

		if err != nil {
			continue
		}

		if match {
			return &redirect
		}
	}

	return nil
}

func getHeaderRules(deployId int64) (*[]HeaderRule, error) {
	return cache.UseCache(
		fmt.Sprintf("header__%d", deployId),
		[]cache.ResolverSetter[[]HeaderRule]{
			cache.UseLocalCache[[]HeaderRule](),
			cache.UseRedisCache[[]HeaderRule](),
			{
				Resolver: func(key string) (*[]HeaderRule, error) {
					session := db.Get()
					data := models.DeploymentConfigsStruct{
						DeployId: deployId,
					}

					q := session.Query(models.DeploymentConfigs.Get("headers")).BindStruct(data)
					err := q.GetRelease(&data)
					log.Printf("getHeaderRulesDB: %v", data)

					result := make([]HeaderRule, 0)

					if err = json.Unmarshal([]byte(data.Headers), &result); err != nil {
						return nil, err
					}

					return &result, err
				},
			},
		},
	)
}

func MatchHeaders(headers *[]HeaderRule, path string) *HeaderRule {
	if headers == nil {
		return nil
	}

	for _, header := range *headers {
		match, err := regexp.MatchString(header.Pattern, path)

		if err != nil {
			continue
		}

		if match {
			return &header
		}
	}

	return nil
}

func getRewriteRules(deployId int64) (*[]RewriteRule, error) {
	return cache.UseCache(
		fmt.Sprintf("rewrite_%d", deployId),
		[]cache.ResolverSetter[[]RewriteRule]{
			cache.UseLocalCache[[]RewriteRule](),
			cache.UseRedisCache[[]RewriteRule](),
			{
				Resolver: func(key string) (*[]RewriteRule, error) {
					session := db.Get()
					data := models.DeploymentConfigsStruct{
						DeployId: deployId,
					}

					q := session.Query(models.DeploymentConfigs.Get("rewrites")).BindStruct(data)
					err := q.GetRelease(&data)
					log.Printf("getRewriteRulesDB: %v", data)

					result := make([]RewriteRule, 0)

					if err = json.Unmarshal([]byte(data.Rewrites), &result); err != nil {
						return nil, err
					}

					return &result, err
				},
			},
		},
	)
}

func MatchRewrites(rewrites *[]RewriteRule, path string) *RewriteRule {
	if rewrites == nil {
		return nil
	}

	for _, rewrite := range *rewrites {
		match, err := regexp.MatchString(rewrite.Pattern, path)

		if err != nil {
			continue
		}

		if match {
			return &rewrite
		}
	}

	return nil
}

func GetHrrRules(deployID int64) (*[]HeaderRule, *[]RedirectRule, *[]RewriteRule) {
	var headerRules *[]HeaderRule
	var redirectRules *[]RedirectRule
	var rewriteRules *[]RewriteRule

	wg := sync.WaitGroup{}
	wg.Add(3)

	//	Header Rules
	//	TODO - implement
	go func() {
		defer wg.Done()

		rules, err := getHeaderRules(deployID)
		if err != nil {
			log.Printf("GetHeaderRulesErr: %v", err)
		}
		headerRules = rules
	}()

	//	Redirect Rules
	go func() {
		defer wg.Done()

		rules, err := getRedirectRules(deployID)
		if err != nil {
			log.Printf("GetRedirectRulesErr: %v", err)
		}
		redirectRules = rules
	}()

	//	Rewrite Rules
	//	TODO - implement
	go func() {
		defer wg.Done()

		rules, err := getRewriteRules(deployID)
		if err != nil {
			log.Printf("GetRewriteRulesErr: %v", err)
		}
		rewriteRules = rules
	}()

	wg.Wait()

	return headerRules, redirectRules, rewriteRules
}
